File: robin_scalper/core/state.py
"""
策略运行状态（实时状态，存于内存，JSON 序列化用于 Web 推送）。
"""
from __future__ import annotations
from dataclasses import dataclass, field, asdict
from typing import List, Optional, Dict, Any
import time


@dataclass
class GridLeg:
    index: int
    side: str          # "BUY" / "SELL"
    entry_price: float
    lot_size: float
    open_time: float

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class SideState:
    active: bool = False
    avg_price: float = 0.0
    trail_level: float = 0.0
    sl_set: bool = False
    level1_done: bool = False
    level2_done: bool = False
    breakeven_triggered: bool = False
    grids: List[GridLeg] = field(default_factory=list)

    def reset(self) -> None:
        self.active = False
        self.avg_price = 0.0
        self.trail_level = 0.0
        self.sl_set = False
        self.level1_done = False
        self.level2_done = False
        self.breakeven_triggered = False
        self.grids = []


@dataclass
class EngineState:
    # 引擎
    running: bool = False
    paper_trading: bool = True
    last_tick_ts: float = 0.0
    last_eval_text: str = ""
    last_eval_time: float = 0.0
    # 价格
    bid: float = 0.0
    ask: float = 0.0
    last_price: float = 0.0
    # 指标
    rsi: float = 0.0
    htf_trend: str = ""  # "多" / "空" / "震荡" / ""
    htf_price: float = 0.0
    htf_ma1: float = 0.0
    htf_ma2: float = 0.0
    htf_adx: float = 0.0
    htf_macd_main: float = 0.0
    htf_macd_signal: float = 0.0
    # 风控
    daily_open_count: int = 0
    last_open_date: int = 0
    last_close_time: float = 0.0
    last_maxloss_time: float = 0.0
    # 状态
    buy: SideState = field(default_factory=SideState)
    sell: SideState = field(default_factory=SideState)
    # 风控触发
    in_no_trade_time: bool = False
    daily_limit_reached: bool = False
    wait_after_close: bool = False
    wait_after_maxloss: bool = False
    # 持仓汇总
    total_long_lots: float = 0.0
    total_short_lots: float = 0.0
    unrealized_pnl: float = 0.0
    # 错误
    last_error: str = ""
    # 延迟
    ws_latency_ms: float = 0.0  # 最后一次延迟（ms）

    def snapshot(self) -> dict:
        d = asdict(self)
        return d

    def reset_side(self, side: str) -> None:
        if side == "BUY":
            self.buy.reset()
        else:
            self.sell.reset()

File: robin-client/src/core/state.ts
// 策略运行状态 - 对应 Python 版本的 state.py

import { GridLeg, SideState, LogEntry, TickData } from './types';

export class EngineState {
  // 引擎状态
  running = false;
  paper_trading = true;
  last_tick_ts = 0;
  last_eval_text = '';
  last_eval_time = 0;

  // 价格
  bid = 0;
  ask = 0;
  last_price = 0;

  // 指标
  rsi = 0;
  htf_trend: '多' | '空' | '震荡' | '' = '';
  htf_price = 0;
  htf_ma1 = 0;
  htf_ma2 = 0;
  htf_adx = 0;
  htf_macd_main = 0;
  htf_macd_signal = 0;

  // 风控
  daily_open_count = 0;
  last_open_date = 0;
  last_close_time = 0;
  last_maxloss_time = 0;

  // 多空状态
  buy: SideState = {
    active: false,
    avg_price: 0,
    trail_level: 0,
    sl_set: false,
    level1_done: false,
    level2_done: false,
    breakeven_triggered: false,
    grids: [],
  };

  sell: SideState = {
    active: false,
    avg_price: 0,
    trail_level: 0,
    sl_set: false,
    level1_done: false,
    level2_done: false,
    breakeven_triggered: false,
    grids: [],
  };

  // 风控触发标志
  in_no_trade_time = false;
  daily_limit_reached = false;
  wait_after_close = false;
  wait_after_maxloss = false;

  // 持仓汇总
  total_long_lots = 0;
  total_short_lots = 0;
  unrealized_pnl = 0;

  // 错误
  last_error = '';

  // 延迟
  ws_latency_ms = 0;

  snapshot(): any {
    return {
      running: this.running,
      paper_trading: this.paper_trading,
      last_tick_ts: this.last_tick_ts,
      last_eval_text: this.last_eval_text,
      last_eval_time: this.last_eval_time,
      bid: this.bid,
      ask: this.ask,
      last_price: this.last_price,
      rsi: this.rsi,
      htf_trend: this.htf_trend,
      htf_price: this.htf_price,
      htf_ma1: this.htf_ma1,
      htf_ma2: this.htf_ma2,
      htf_adx: this.htf_adx,
      htf_macd_main: this.htf_macd_main,
      htf_macd_signal: this.htf_macd_signal,
      daily_open_count: this.daily_open_count,
      last_open_date: this.last_open_date,
      last_close_time: this.last_close_time,
      last_maxloss_time: this.last_maxloss_time,
      buy: { ...this.buy },
      sell: { ...this.sell },
      in_no_trade_time: this.in_no_trade_time,
      daily_limit_reached: this.daily_limit_reached,
      wait_after_close: this.wait_after_close,
      wait_after_maxloss: this.wait_after_maxloss,
      total_long_lots: this.total_long_lots,
      total_short_lots: this.total_short_lots,
      unrealized_pnl: this.unrealized_pnl,
      last_error: this.last_error,
      ws_latency_ms: this.ws_latency_ms,
    };
  }

  resetSide(side: 'BUY' | 'SELL'): void {
    if (side === 'BUY') {
      this.buy = {
        active: false,
        avg_price: 0,
        trail_level: 0,
        sl_set: false,
        level1_done: false,
        level2_done: false,
        breakeven_triggered: false,
        grids: [],
      };
    } else {
      this.sell = {
        active: false,
        avg_price: 0,
        trail_level: 0,
        sl_set: false,
        level1_done: false,
        level2_done: false,
        breakeven_triggered: false,
        grids: [],
      };
    }
  }
}
